apis/Google.Cloud.Metadata.V1/Google.Cloud.Metadata.V1/MetadataClientImpl.cs

```diff
@@ -0,0 +1,502 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Apis.Auth.OAuth2.Responses;
+using Google.Apis.Compute.v1.Data;
+using Google.Apis.Http;
+using Google.Apis.Json;
+using Google.Apis.Logging;
+using Microsoft.CSharp.RuntimeBinder;
+using Newtonsoft.Json.Linq;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Net.Http;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using static Google.Apis.Compute.v1.Data.Metadata;
+
+namespace Google.Cloud.Metadata.V1
+{
+    /// <summary>
+    /// Wrapper around metadata server requests to provide simpler operations.
+    /// </summary>
+    /// <remarks>
+    /// This is the "default" implementation of <see cref="MetadataClient"/>. Most client code
+    /// should refer to <see cref="MetadataClient"/>, creating instances with
+    /// <see cref="MetadataClient.Create(ConfigurableHttpClient)"/> and
+    /// <see cref="MetadataClient.CreateAsync(IHttpClientFactory)"/>. The constructor
+    /// of this class is public for the sake of constructor-based dependency injection.
+    /// </remarks>
+    public sealed class MetadataClientImpl : MetadataClient
+    {
+        private const string DefaultMetadataHost = "metadata";
+        private const string EmulatorEnvironmentVariable = "METADATA_EMULATOR_HOST";
+        private const string MetadataFlavor = "Metadata-Flavor";
+        private const string GoogleMetadataHeader = "Google";
+
+        // TODO: Should this be created with a MetadataClient instead?
+        private static readonly ILogger Logger = ApplicationContext.Logger.ForType<MetadataClientImpl>();
+
+        private readonly string metadataHost;
+
+        private readonly object eventLock = new object();
+        private ValueWatcher instanceWatcher;
+        private ValueWatcher maintenanceStatusWatcher;
+        private ValueWatcher projectWatcher;
+
+        public override ConfigurableHttpClient HttpClient { get; }
+
+        /// <summary>
+        /// Constructs a new client for the metadata server.
+        /// </summary>
+        /// <param name="httpClient">The <see cref="ConfigurableHttpClient"/> to use when making requests, or null to create a default instance.</param>
+        public MetadataClientImpl(ConfigurableHttpClient httpClient = null)
+        {
+            HttpClient = httpClient ?? new HttpClientFactory().CreateHttpClient(new CreateHttpClientArgs());
```

Comment:
This statement is a bit too complex to parse for my taste, can we split it in two variables for readability. Is the two `new` operators that confuse me.